hal/reset_reason_api.h

```diff
@@ -0,0 +1,110 @@
+ * limitations under the License.
+ */
+#ifndef MBED_RESET_REASON_API_H
+#define MBED_RESET_REASON_API_H
+
+#if DEVICE_RESET_REASON
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_reset_reason Reset Reason HAL API
+ * @{
+ */
+
+/** Definitions of different reset reasons
+ */
+typedef enum {
+    RESET_REASON_POWER_ON,       /**< Set when power is initially applied to the board. The power-on-reset circuit causes a POWER_ON reset when this occurs */
+    RESET_REASON_PIN_RESET,      /**< Set when a reset is triggered by the hardware pin on the board */
+    RESET_REASON_BROWN_OUT,      /**< Triggered when the voltage drops below the low voltage detect (LVD) threshold the system will be held in a reset until the voltage rises above the threshold */
+    RESET_REASON_SOFTWARE,       /**< Set during software reset, typically triggered by writing the SYSRESETREQ bit in the Application Interrupt and Reset Control register */
+    RESET_REASON_WATCHDOG,       /**< Set when a running watchdog timer fails to be refreshed */
+    RESET_REASON_LOCKUP,         /**< Set when the core is locked because of an unrecoverable exception */
+    RESET_REASON_WAKE_LOW_POWER, /**< Set when waking from deep sleep mode */
+    RESET_REASON_ACCESS_ERROR,   /**< Umbrella value that encompasses any access related reset */
+    RESET_REASON_BOOT_ERROR,     /**< Umbrella value that encompasses any boot related reset */
+    RESET_REASON_MULTIPLE,       /**< Set if multiple reset reasons are set within the board. Occurs when the reset reason registers aren't cleared between resets */
+    RESET_REASON_PLATFORM,       /**< Platform specific reset reason not captured in this enum */
+    RESET_REASON_UNKNOWN         /**< Unknown or unreadable reset reason **/
+} reset_reason_t;
+
+/** Fetch the reset reason for the last system reset
+ *
+ * This function must return the contents of the system reset reason registers
+ * cast to an appropriate platform independent reset reason. If multiple reset
+ * reasons are set this function should return RESET_REASON_MULTIPLE. If the
+ * reset reason does not match any existing platform independent value this
+ * function should return RESET_REASON_PLATFORM. If no reset reason can be
+ * determined this function should return RESET_REASON_UNKNOWN.
+ *
+ * This function is not idempotent, there is no guarantee that the system
+ * reset reason will not be cleared between calls to this function altering the
+ * return value between calls.
+ *
+ * Note: Some platforms contain reset reason registers that persist through
+ * system resets. If the registers haven't been cleared before calling this
+ * function multiple reasons may be set within the registers. If multiple reset
+ * reasons are detected this function will return RESET_REASON_MULTIPLE.
+ *
+ * @return enum containing the last reset reason for the board.
+ */
+reset_reason_t hal_reset_reason_get(void);
```

Comment:
This may be something that needs to be addressed at the driver layer, but what are the guidelines for using this? If the reset flags are cleared on first read, then applications with a bootloader, like cloud client, may never know the reset reason if the bootloader uses the watchdog.

Should there be a recommendation like this `shouldn't be used from bootloaders` or `must always be called on boot from the main application`? Alternatively should this be handled at the driver layer and stored into persistent ram, so bootloader and applications are free to use the driver API? Previous discussion [here](https://github.com/ARMmbed/mbed-os/pull/5488#discussion_r153566734). @SenRamakri @deepikabhavnani what are your thoughts on this?

Reply:
Agree with @c1728p9 - Reset reason should not be cleared by HAL, also if it is cleared by underlying hardware (clear on read) should be taken care by driver. Drivers are thread safe, so makes more sense in having storage in driver